responsemanager/responseassembler/responseassembler.go

```diff
@@ -70,46 +64,99 @@ func New(ctx context.Context, peerHandler PeerMessageHandler) *ResponseAssembler
+	return &responseStream{
+		requestID:      requestID,
+		p:              p,
+		messageSenders: ra.peerHandler,
+		linkTrackers:   ra.PeerManager,
+		subscriber:     subscriber,
+	}
+}
+
+type responseStream struct {
+	requestID      graphsync.RequestID
+	p              peer.ID
+	closed         bool
+	closedLk       sync.RWMutex
+	messageSenders PeerMessageHandler
+	linkTrackers   *peermanager.PeerManager
+	subscriber     notifications.Subscriber
+}
+
+func (r *responseStream) Close() error {
+	r.closedLk.Lock()
+	r.closed = true
+	r.closedLk.Unlock()
+	return nil
+}
+
+func (r *responseStream) isClosed() bool {
+	r.closedLk.RLock()
+	defer r.closedLk.RUnlock()
+	return r.closed
+}
+
+type ResponseStream interface {
+	Transaction(transaction Transaction) error
+	DedupKey(key string)
+	IgnoreBlocks(links []ipld.Link)
+	SkipFirstBlocks(skipFirstBlocks int64)
+	// ClearRequest removes all tracking for this request.
+	ClearRequest()
+}
+
 // DedupKey indicates that outgoing blocks should be deduplicated in a seperate bucket (only with requests that share
 // supplied key string)
-func (ra *ResponseAssembler) DedupKey(p peer.ID, requestID graphsync.RequestID, key string) {
-	ra.GetProcess(p).(*peerLinkTracker).DedupKey(requestID, key)
+func (rs *responseStream) DedupKey(key string) {
+	rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).DedupKey(rs.requestID, key)
 }
 
 // IgnoreBlocks indicates that a list of keys should be ignored when sending blocks
-func (ra *ResponseAssembler) IgnoreBlocks(p peer.ID, requestID graphsync.RequestID, links []ipld.Link) {
-	ra.GetProcess(p).(*peerLinkTracker).IgnoreBlocks(requestID, links)
+func (rs *responseStream) IgnoreBlocks(links []ipld.Link) {
+	rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).IgnoreBlocks(rs.requestID, links)
 }
 
 // SkipFirstBlocks tells the assembler for the given request to not send the first N blocks
-func (ra *ResponseAssembler) SkipFirstBlocks(p peer.ID, requestID graphsync.RequestID, skipFirstBlocks int64) {
-	ra.GetProcess(p).(*peerLinkTracker).SkipFirstBlocks(requestID, skipFirstBlocks)
+func (rs *responseStream) SkipFirstBlocks(skipFirstBlocks int64) {
+	rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).SkipFirstBlocks(rs.requestID, skipFirstBlocks)
+}
+
+// ClearRequest removes all tracking for this request.
+func (rs *responseStream) ClearRequest() {
+	_ = rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).FinishTracking(rs.requestID)
 }
 
-// Transaction builds a response, and queues it for sending in the next outgoing message
-func (ra *ResponseAssembler) Transaction(p peer.ID, requestID graphsync.RequestID, transaction Transaction) error {
+func (rs *responseStream) Transaction(transaction Transaction) error {
 	rb := &responseBuilder{
-		requestID:   requestID,
-		linkTracker: ra.GetProcess(p).(*peerLinkTracker),
+		requestID:   rs.requestID,
+		linkTracker: rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker),
 	}
 	err := transaction(rb)
```

Comment:
Do we need to run this if `rs.isClosed() == true`? Could we also be short-circuiting in this method as well?

Reply:
possibly... but I'm a little worried we still have logic unrelated to whats going over the wire happening inside the transaction, and I don't want to short circuit that.